core/src/main/scala/kafka/server/ReplicaFetcherThread.scala

```diff
@@ -254,15 +255,30 @@ class ReplicaFetcherThread(name: String,
       // We will not include a replica in the fetch request if it should be throttled.
       if (fetchState.isReadyForFetch && !shouldFollowerThrottle(quota, fetchState, topicPartition)) {
         try {
-          val logStartOffset = this.logStartOffset(topicPartition)
-          builder.add(topicPartition, new FetchRequest.PartitionData(
-            fetchState.fetchOffset, logStartOffset, fetchSize, Optional.of(fetchState.currentLeaderEpoch)))
+          // A fetch needs to be used to update the session only if:
+          //   1. fetch offset has changed
+          //   2. log start offset has changed
+          //   3. leader epoch has changed
+          //   4. the fetch was excluded before and now
+          if (fetchState.updateNeeded) {
+            val logStartOffset = this.logStartOffset(topicPartition)
```

Comment:
The logStartOffset may change due to actions in the Partition/Log which are outside of the replica fetcher and are not currently signalled to the replica fetcher e.g. retention. I believe we need a way for the Partition/Log to signal that an update has occurred.

Reply:
I was wondering whether this was possible, but I really don't have a good handle with whether it's safe or not. Syncing with others to make sure makes sense.

Reply:
Synced with @junrao @hachikuji offline, and I think the log-start-offset is used for KIP-107 only, so not updating it due to log retention when sending to leader is okay.